src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,769 @@
+                                                queryContinuation.Identifier,
+                                                argumentList: null,
+                                                SyntaxFactory.EqualsValueClause(selectClause.Expression))))), s));
+
+                        return TryProcessQueryBody(queryContinuation.Body, processingState);
+                    }
+                }
+
+                return false;
+            }
+
+            private bool TryProcessWhereClause(WhereClauseSyntax whereClause, ProcessingState processingState)
+            {
+                if (IsValidWhereClauseSymbol(whereClause))
+                {
+                    processingState.AddStatementProcessingMethod(
+                         s => SyntaxFactory.Block(
+                             SyntaxFactory.IfStatement(
+                                 ProcessCondition((whereClause).Condition),
+                                 s)));
+
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool IsValidWhereClauseSymbol(WhereClauseSyntax whereClause)
+            {
+                if (!TryGetMethodSymbol(whereClause, out var methodSymbol))
+                {
+                    return false;
+                }
+
+                var returnType = methodSymbol.ReturnType;
+                if (returnType.Name != "IEnumerable") // TODO better that name check?
+                {
+                    return false;
+                }
+
+                var typeArguments = returnType.GetTypeArguments();
+                if (typeArguments.Length != 1)
+                {
+                    return false;
+                }
+
+                var typeArgument = typeArguments.Single();
+                switch (methodSymbol.Parameters.Length)
+                {
+                    // IEnumerable<T> Where(Func<T, bool> predicate) 
+                    case 1:
+                        return IsPredicate(methodSymbol.Parameters.First().Type, typeArgument);
+
+                    // IEnumerable<T> Where(this IEnumerable<T>, Func<T, bool> predicate) 
+                    case 2:
+                        // TODO add checks for this.
+                        return IsPredicate(methodSymbol.Parameters.First().Type, typeArgument);
+
+                    default: return false;
+                }
+            }
+
+            private bool TryGetMethodSymbol(QueryClauseSyntax queryClause, out IMethodSymbol methodSymbol)
+            {
+                QueryClauseInfo queryClauseInfo = _semanticModel.GetQueryClauseInfo(queryClause, _cancellationToken);
+                methodSymbol = queryClauseInfo.OperationInfo.Symbol as IMethodSymbol;
+                return methodSymbol != null;
+            }
+
+            private static bool IsPredicate(ITypeSymbol typeSymbol, ITypeSymbol typeArgument)
+            {
+                if (typeSymbol.Name != "Func")  // TODO better that name check?
```

Comment:
yes.  you can grab ```System.Func<X,Y>``` from the compilation you have, using ```GetTypeByMetataName(typeof(System.Func<,>).FullName```.   Then compare typeSymbol.OriginalDefinition to that symbol.

Reply:
Is not using compilation for the purpose expensive?
Do you mean something like
document.Project.GetCompilationAsync().GetAssemblyOrModuleSymbol(reference).GetTypeByMetataName()
?

---
In reply to: [176869034](https://github.com/dotnet/roslyn/pull/25362#discussion_r176869034) [](ancestors = 176869034)

Reply:
you are already using symbols. so it should be fine. if you only uses syntax node and use compilation only for this, then, yes. but in this case, you are already doing all kind of binding, symbol walking, so shouldn't matter.

Reply:
Do you mean that '(typeSymbol.Name != "Func")' is the right check here?

---
In reply to: [176895380](https://github.com/dotnet/roslyn/pull/25362#discussion_r176895380) [](ancestors = 176895380)

Reply:
no I mean was what cyrus suggested is right way. 

I was answering on your concern.
> Is not using compilation for the purpose expensive?

var funcSymbol = compilation.GetTypeByMetadataName(typeof(...).FullName)

Reply:
Thank you! So, do I have a compilation in the context or should I create one?

---
In reply to: [176896675](https://github.com/dotnet/roslyn/pull/25362#discussion_r176896675) [](ancestors = 176896675)

Reply:
you can get compilation from document. but if you already have semanticmodel, you can get one from it as well. semanticModel.Compilation.